subcommand/server-acl-init/command_test.go

```diff
@@ -1312,6 +1488,156 @@ func TestRun_HTTPS(t *testing.T) {
 	require.True(ok)
 }
 
+// Test that the ACL replication token created from the primary DC can be used
+// for replication in the secondary DC.
+func TestRun_ACLReplicationTokenValid(t *testing.T) {
+	t.Parallel()
+
+	// Set up the primary DC first.
+	primarySvr, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.ACL.Enabled = true
+	})
+	require.NoError(t, err)
+	defer primarySvr.Stop()
+
+	primaryK8s := fake.NewSimpleClientset()
+	createTestK8SResources(t, primaryK8s, primarySvr.HTTPAddr, resourcePrefix, "http", ns)
+	require.NoError(t, err)
+
+	// Run the command to bootstrap ACLs
+	primaryUI := cli.NewMockUi()
+	primaryCmd := Command{
+		UI:        primaryUI,
+		clientset: primaryK8s,
+	}
+	primaryCmd.init()
+	primaryCmdArgs := []string{
+		"-k8s-namespace=" + ns,
+		"-expected-replicas=1",
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-create-acl-replication-token",
+	}
+	responseCode := primaryCmd.Run(primaryCmdArgs)
+	require.Equal(t, 0, responseCode, primaryUI.ErrorWriter.String())
+
+	// Retrieve the replication ACL token from the kubernetes secret.
+	tokenSecret, err := primaryK8s.CoreV1().Secrets(ns).Get("release-name-consul-acl-replication-acl-token", metav1.GetOptions{})
+	require.NoError(t, err)
+	require.NotNil(t, tokenSecret)
+	aclReplicationTokenBytes, ok := tokenSecret.Data["token"]
+	require.True(t, ok)
+	aclReplicationToken := string(aclReplicationTokenBytes)
+
+	// Now we can set up the secondary DC.
+	secondarySvr, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.Datacenter = "dc2"
+		c.ACL.Enabled = true
+		c.ACL.TokenReplication = true
+		c.PrimaryDatacenter = "dc1"
+		// When running via the Helm chart these are passed in to the servers.
+		c.ACL.Tokens.Agent = aclReplicationToken
+		c.ACL.Tokens.Replication = aclReplicationToken
+	})
+	require.NoError(t, err)
+	defer secondarySvr.Stop()
+
+	// When running via Helm, the WAN join config is set in the server config
+	// however that config isn't exposed by the testutil server so instead we need to
+	// make the join API call. This requires the bootstrap token permissions.
+	bootToken := getBootToken(t, primaryK8s, resourcePrefix, ns)
+	secondaryConsulClient, err := api.NewClient(&api.Config{
+		Address: secondarySvr.HTTPAddr,
+		Token:   bootToken,
+	})
+	err = secondaryConsulClient.Agent().Join(primarySvr.WANAddr, true)
+	require.NoError(t, err)
```

Comment:
It's similar setup but different in some important ways. I decided to keep the code separate because I didn't want to overly complicate the `completeReplicatedSetup` code. I'll note the differences and then you can decide if it makes sense to combine them.

Differences:
1. In this test, ACLs in the primary DC are bootstrapped via the server-acl-init command actually being run. In `completeReplicatedSetup` we set the agent master token on the server before startup so we can skip acl bootstrap. This saves us from testing out the server-acl-init command in `completeReplicatedSetup` because what we really want to test is the secondary DC. In this test, we want to test the whole flow so we run the full command.
  
    To combine them, we could either always run the server-acl-init command in `completeReplicatedSetup` or we could add a conditional to the function to select whether we run the command or not.
1. In this test, the secondary DC's replication token is the one created from running the server-acl-init command in the first DC. In `completeReplicatedSetup` we use the bootToken as the replication token because it saves us having to create the policy and acl token in the primary DC.
1. In this test, we need access to the replicationToken. We'd need to return it from `completeReplicatedSetup`.
1. In this test, we don't need to pre-specify a bootToken since that automatically gets created when we run server-acl-init so we'd end up passing an empty string to `completeReplicatedSetup`

Here's an idea of what the change could look like: https://github.com/hashicorp/consul-k8s/commit/5f122d8cfebba90bd2d7688d9b9061ce6b53f657

Reply:
Thanks for including the code sample. I like how it looks. The setup function doesn't seem too complex to me and I like that the test looks much simpler and easier to read. So if I were to decide, I'd go for this refactor.

Reply:
done.